src/Mod/Part/Gui/SectionCutting.cpp

```diff
@@ -137,16 +137,65 @@ SectionCut::SectionCut(QWidget* parent)
-        ui->AutoCutfaceColor->setChecked(false);
+        ui->autoCutfaceColorCB->setChecked(false);
+        ui->autoBFColorCB->setChecked(false);
         if (doc->getObject(CompoundName)) {
+            // get the object with the right name
             auto compoundObject = doc->getObject(CompoundName);
-            Part::Compound* pcCompound = dynamic_cast<Part::Compound*>(compoundObject);
-            if (!pcCompound) {
-                Base::Console().Error(
-                    "SectionCut error: compound is incorrectly named, cannot proceed\n");
-                return;
+            // to later store the childs
+            std::vector<App::DocumentObject*> compoundChilds;
+
+            // check if this is a BooleanFragments or a Part::Compound
+            // Part::Compound is the case when there was only one object
+            Part::Compound* pcCompoundPart = dynamic_cast<Part::Compound*>(compoundObject);
+            if (!pcCompoundPart) {
+                // for more security check for validity accessing its ViewProvider
+                auto pcCompoundBF = Gui::Application::Instance->getViewProvider(compoundObject);
+                if (!pcCompoundBF) {
+                    Base::Console().Error(
+                        "SectionCut error: compound is incorrectly named, cannot proceed\n");
+                    return;
+                }
+                auto property = compoundObject->getPropertyByName("Shape");
+                Part::PropertyPartShape* objectShape =
+                    static_cast<Part::PropertyPartShape*>(property);
+                BoundCompound = objectShape->getBoundingBox();
+                // for BooleanFragments we also need to set the checkbox, transparency and color
+                ui->groupBoxIntersecting->setChecked(true);
+                auto pcCompoundBFGO = dynamic_cast<Gui::ViewProviderGeometryObject*>(pcCompoundBF);
+                if (pcCompoundBFGO) {
+                    App::Color compoundColor = pcCompoundBFGO->ShapeColor.getValue();
+                    ui->BFragColor->setColor(compoundColor.asValue<QColor>());
+                    int compoundTransparency = pcCompoundBFGO->Transparency.getValue();
+                    ui->BFragTransparencyHS->setValue(compoundTransparency);
+                    ui->BFragTransparencyHS->setToolTip(QString::number(compoundTransparency)
+                                                        + QString::fromLatin1(" %"));
+                    // Part::Cut ignores the cutbox transparency when it is set
+                    // to zero and the BooleanFragments transparency is not zero
+                    // therefore limit the cutbox transparency to 1 in this case
+                    if (compoundTransparency > 0)
+                        ui->CutTransparencyHS->setMinimum(1);
+                    else
+                        ui->CutTransparencyHS->setMinimum(0);
+                }
+                compoundChilds = pcCompoundBF->claimChildren();

```

Comment:
The CI complaint about cognitive complexity is primarily due to the deeply-nested if statements. You could consider splitting this block into its own function, which would probably about cut the complexity in half.

Reply:
Regarding where I see it: when you view the "Files" section of the PR, the CI's checker warnings display inline in the code. They are only warnings, so don't cause the checks to fail.

Reply:
> I have not seen any modern software design advice advocating for **ten** nesting levels. Are you sure you aren't seeing references to 10 conditionals, rather than 10 nested levels?

They are not advocating it, say one should keep it low and not exceed more than 10. (which is obvious since this is then no longer to get by a human)

> It certainly improves the code's _testability_, but I am less convinced that it is improving its _understandability_. But perhaps it is, I'm too close to the code to tell for sure.

This is also my issue. I try to code so that others but also myself can understand what is done in e.g. 2 years. Therefore I use a lot of comments and split to functions and lambdas for repeated tasks.
I had a look and here i would keep the if structure because the logic is not trivial and by having it in a single procedure keeps the reading flow to understand what is done and why. I fear that when i sort out some ifs, another one (and myself in future of course), has to scroll around to see what is done in the functions that are called. This is not helpful to get the logic.
So for me the question is what would be the benefit. The code execution is not quicker, memory space is not saved, it is a bit harder to understand the code logic.
However, I might be too deep into it at the moment. I rely on your judgement here.

Reply:
> Regarding where I see it: when you view the "Files" section of the PR, the CI's checker warnings display inline in the code. They are only warnings, so don't cause the checks to fail.

So you mean GitHub Actions? The report is
https://clang.llvm.org/extra/clang-tidy/checks/readability/function-cognitive-complexity.html
I never heard of this yet. To me the limit of 25 is rather arbitrary. As a physicist I always ask myself why not 24 or 26. If there is no good explanation, I am sceptical.
Sure the nesting level should not be to great but nesting is hereby just one factor. The besting depth is in my case 5, but the main cutting void is large. It is, but it covers the whole cutting process. My idea was hereby that I can follow to read this void and can step through the whole process and understand it. Now that I haven't touched the cutting code for months, it helped me to find the issues. I stepped through the code and this way could identify the position in the process where an issue occurs/is created.
But again, I am the author of that code, what is great for me, can for others be a nightmare. That's why I will rely on your judgement.

However, in the CMake build report of the CI I see 
`warning: unused variable ‘compoundTransparency’`
so there is a bug in my code. I will have a look.

Reply:
> so there is a bug in my code. I will have a look.

Now fixed. I also modernized now all for loops

Reply:
Certainly the 25 limit is arbitrary -- so is the definition of "complexity" itself. Linters like this are meant to be advice, not a hard-and-fast rule. Clean Code would argue that a complexity of 25 is itself absurdly high, but I bet a linter run on all of FreeCAD yields thousands of functions over that limit. Which is to say: this code is certainly no worse than other vast swaths of our legacy codebase, and until we are ready to start adding unit tests for real, it's sort of academic. Break it up or not, your call. That said, if you ever have free time for some "light" reading, Clean Code is a very highly regarded treatise on the subject, even if you don't follow Agile/TDD/etc.

Reply:
OK. For now and for the given reasons I would keep the structure as it is. One this has been merged, I expect that there are still issues under special circumstances and I would like to resolve them before I refactor the code. I expect the issues because SectionCutting is complex in terms of a high possible number of object combinations (normal objects, links, parts, part childs, links of links etc.

Reply:
>  I fear that when i sort out some ifs, another one (and myself in future of course), has to scroll around to see what > is done in the functions that are called. This is not helpful to get the logic.

As Uncle Bob has already been mentioned in this context. One of the aspects of clean code and SRP is that a function should do exactly one thing. Mixing levels of abstraction inside a function should be avoided because it's confusing as you have to keep in mind the higher level logic and at the same time get right the lower level logic. By splitting this into several small functions or even classes you can separate the different levels of abstractions.
The function names should be carefully chosen so that just by reading the name a reader already knows what it is supposed to do. Don't hesitate to use lengthy function names.

> So for me the question is what would be the benefit. The code execution is not quicker, memory space is not saved, it is a bit harder to understand the code logic.

Forget about execution speed or memory. Compilers nowadays are smart enough to optimize the assembler code better than an average programmer could do. The idea is to split the entire complexity into smaller less complex parts and the advantage is that the code becomes much more readable, maintainable and extendable. And of course it will be much easier to write test functions.

Things might sound too abstract to you. So, I give you a concrete example of some refactoring I did a few weeks ago. The implementation of CmdPartDesignNewSketch in the past was this big mess:
https://github.com/FreeCAD/FreeCAD/blob/674bd3d6ec59a46dbcfa2f9891eb085e12976e44/src/Mod/PartDesign/Gui/Command.cpp#L494

After doing some iterations of code refactoring it has become a two-liner:
```
PartDesignGui::SketchWorkflow creator(getActiveGuiDocument());
creator.createSketch();
```

The whole logic has been moved to a new class SketchWorkflow and the entry function is createSketch() -- https://github.com/FreeCAD/FreeCAD/blob/master/src/Mod/PartDesign/Gui/SketchWorkflow.cpp#L642

Reply:
> Forget about execution speed or memory. Compilers nowadays are smart enough to optimize the assembler code better than an average programmer could do. The idea is to split the entire complexity into smaller less complex parts and the advantage is that the code becomes much more readable, maintainable and extendable. And of course it will be much easier to write test functions.

As I wrote, I am open for a revision. However, at first, I would like to have the PR before I start a revision. This would be much easier for me.
Besides this, the SectionCutting code is now in for a while and you once approved it. It was a month-long process of getting it ready to be merged. In that time I could have written it in the way you prefer ;-).

I think that I take care of that my code is understandable. It is a personal preference that I like to follow a logic line by line without the need to scroll around to see what the different called functions do. For repeated tasks I already use lambdas whenever possible. When this is not suitable to be able to write test functions, then I need to learn how long a void can be.